.github/workflows/ci.yml

```diff
@@ -1,32 +1,100 @@
 name: CI
 
-env:
-  COMPOSE_INTERACTIVE_NO_CLI: 1
-  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-
 on:
   push:
-    branches: [ master ]
   pull_request:
-    branches: [ master ]
+    branches: [ 3.x, 2.x, master ]
+env:
+  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
 
 jobs:
   tests:
     runs-on: ubuntu-latest
+    container: abrardev/tenancy:latest
 
     strategy:
       matrix:
-        php: ["8.1"]
-        laravel: ["^9.0"]
+        laravel: ['^8.0', '^9.0']
+
+    services:
+      postgres:
+        image: postgres:latest
+        env:
+          POSTGRES_PASSWORD: password
+          POSTGRES_USER: root
+          POSTGRES_DB: main
+        ports:
+          - 5432/tcp
+        options: --health-cmd pg_isready --health-interval 10s --health-timeout 5s --health-retries 3
+
+      mysql:
+        image: mysql:5.7
+        env:
+          MYSQL_ALLOW_EMPTY_PASSWORD: false
+          MYSQL_ROOT_PASSWORD: password
+          MYSQL_DATABASE: main
+        ports:
+          - 3306/tcp
+        options: --health-cmd="mysqladmin ping" --health-interval=10s --health-timeout=5s --health-retries=3
+
+      mysql2:
+        image: mysql:5.7
+        env:
+          MYSQL_ALLOW_EMPTY_PASSWORD: false
+          MYSQL_ROOT_PASSWORD: password
+          MYSQL_DATABASE: main
+        ports:
+          - 3306/tcp
+        options: --health-cmd="mysqladmin ping" --health-interval=10s --health-timeout=5s --health-retries=3
+
+      mssql:
+        image: mcr.microsoft.com/mssql/server:2019-latest
+        ports:
+          - 1433/tcp
+        env:
+          ACCEPT_EULA: Y
+          SA_PASSWORD: P@ssword
+        options: --health-cmd "echo quit | /opt/mssql-tools/bin/sqlcmd -S 127.0.0.1 -l 1 -U sa -P P@ssword"
+
+      redis:
+        image: redis
+        ports:
+          - 6379/tcp
+        options: --health-cmd="redis-cli ping" --health-interval=10s --health-timeout=5s --health-retries=3
 
     steps:
-    - uses: actions/checkout@v2
-    - name: Start docker containers
-      run: PHP_VERSION=${{ matrix.php }} docker-compose up -d
-    - name: Install dependencies
-      run: docker-compose exec -T test composer require --no-interaction "laravel/framework:${{ matrix.laravel }}"
-    - name: Run tests
-      run: ./test
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Get composer cache directory
+        id: composer-cache
+        run: echo "::set-output name=dir::$(composer config cache-files-dir)"
+
+      - name: Cache composer dependencies
+        uses: actions/cache@v2
+        with:
+          path: ${{ steps.composer-cache.outputs.dir }}
+          key: ${{ runner.os }}-${{ matrix.laravel }}-${{ hashFiles('**/composer.lock') }}
```

Comment:
Just to confirm: is it tested that this definitely speeds up the CI?

Reply:
I've compared both versions. One with [cached](https://github.com/abrardev99/tenancy/runs/7947399069?check_suite_focus=true) composer dependencies and one [without caching](https://github.com/abrardev99/tenancy/runs/7947643982?check_suite_focus=true). Looks like both are taking the same time even when the cached version is found. Without cache, job finished a few (4-8) seconds early, but sometimes it finished a few seconds later. Not a major difference. 

> is it tested that this definitely speeds up the CI?

I don't think so. Plus, I found this comment from Graham
<img width="911" alt="image" src="https://user-images.githubusercontent.com/54532330/185879009-2bdbb33f-8c85-42bd-b248-66c7c17fe199.png">